File: PokerAlgo/Guards.cs
namespace PokerAlgo;

internal static class Guards
{
    public static void ArgsGetWinners(List<Player> players, List<Card> communityCards)
    {
        if (players.Count < 2) throw new ArgumentOutOfRangeException(nameof(players), "There must be at least 2 players.");
        if (communityCards.Count != 5) throw new ArgumentOutOfRangeException(nameof(communityCards), "For the showdown, there must be all 5 community cards.");

        List<Card> allCards = new();
        allCards.AddRange(communityCards);
        foreach (Player p in players)
        {
            allCards.Add(p.HoleCards.First);
            allCards.Add(p.HoleCards.Second);
        }

        bool areUnique = allCards.Count == allCards.Distinct().Count();
        if (areUnique == false) throw new DuplicateCardException($"Either {nameof(players)} or {nameof(communityCards)} arguments have duplicate cards.");

        foreach (Card c in allCards)
        {
            if (c.Rank == 1) throw new LowAcesException("When instantiating Ace cards use rank 14 no 1.");
        }
    }

    public static void ArgsGetWinningHand(List<Card> cards)
    {
        if (cards.Count < 5 || cards.Count > 7) throw new ArgumentOutOfRangeException(nameof(cards), "The list must have 5-7 cards.");

        bool areUnique = cards.Count == cards.Distinct().Count();
        if (areUnique == false) throw new DuplicateCardException($"{nameof(cards)} argument has duplicate cards.");


        foreach (Card c in cards)
        {
            if (c.Rank == 1) throw new LowAcesException("When instantiating Ace cards use rank 14 not 1.");
        }
    }

    public static void ArgsWinningChanceSim(Pair playerHoleCards, List<Card> communityCards, int numOfOpponents, int numberOfSimulatedGames)
    {
        if (communityCards.Count < 3) throw new ArgumentOutOfRangeException(nameof(communityCards), "There should be no less than 3 community cards.");
        if (communityCards.Count > 5) throw new ArgumentOutOfRangeException(nameof(communityCards), "There should be no more than 5 community cards.");
        if (numOfOpponents < 1) throw new ArgumentOutOfRangeException(nameof(numOfOpponents), "There should be at least 1 opponent.");
        if (numberOfSimulatedGames < 100) throw new ArgumentOutOfRangeException(nameof(numberOfSimulatedGames), "Number of simulated games is less than 100. I recommend at least 100 simulated games for a good prediction.");

        List<Card> allCards = new();
        allCards.AddRange(communityCards);
        allCards.Add(playerHoleCards.First);
        allCards.Add(playerHoleCards.Second);

        bool areUnique = allCards.Count == allCards.Distinct().Count();
        if (areUnique == false) throw new DuplicateCardException($"Either {nameof(playerHoleCards)} or {nameof(communityCards)} arguments have duplicate cards.");
    }

    public static void ArgsPreFlopSim(Pair playerHoleCards, int numOfOpponents, int numberOfSimulatedGames)
    {
        if (numOfOpponents < 1) throw new ArgumentOutOfRangeException(nameof(numOfOpponents), "There should be at least 1 opponent.");
        if (numberOfSimulatedGames < 100) throw new ArgumentOutOfRangeException(nameof(numberOfSimulatedGames), "Number of simulated games is less than 100. I recommend at least 100 simulated games for a good prediction.");

        if (playerHoleCards.First.Rank == playerHoleCards.Second.Rank && playerHoleCards.First.Suit == playerHoleCards.Second.Suit) throw new DuplicateCardException($"{nameof(playerHoleCards)} argument has duplicate cards.");
    }

    public static void ArgsPreFlopLookUp(Pair playerHoleCards, int numOfOpponents)
    {
        if (numOfOpponents < 1) throw new ArgumentOutOfRangeException(nameof(numOfOpponents), "There should be at least 1 opponent.");
        if (playerHoleCards.First.Rank == playerHoleCards.Second.Rank && playerHoleCards.First.Suit == playerHoleCards.Second.Suit) throw new DuplicateCardException($"{nameof(playerHoleCards)} argument has duplicate cards.");
    }

    public static void AgainstDuplicateHoleCards(Pair playerHoleCards, string? paramName)
    {
        if (playerHoleCards.First.Rank == playerHoleCards.Second.Rank && playerHoleCards.First.Suit == playerHoleCards.Second.Suit) throw new DuplicateCardException($"{paramName} argument has duplicate cards.");
    }
    
}

File: guards.go
package pokeralgo

import (
	"fmt"
	"slices"
)

func validatePlayersAndBoard(players []Player, communityCards []Card) error {
	if len(players) < 2 {
		return newError(ErrInvalidArgument, "there must be at least 2 players")
	}
	if len(communityCards) != 5 {
		return newError(ErrInvalidArgument, "for the showdown, there must be all 5 community cards")
	}

	allCards := slices.Clone(communityCards)
	for _, player := range players {
		allCards = append(allCards, player.HoleCards.First, player.HoleCards.Second)
	}

	return validateUniqueCardsAndNoLowAces(allCards, "either players or communityCards arguments have duplicate cards")
}

func validateEvaluationCards(cards []Card) error {
	if len(cards) < 5 || len(cards) > 7 {
		return newError(ErrInvalidArgument, "the list must have 5-7 cards")
	}

	return validateUniqueCardsAndNoLowAces(cards, "cards argument has duplicate cards")
}

func validateSimulation(playerHoleCards HoleCards, communityCards []Card, numOfOpponents int, numberOfSimulatedGames int) error {
	if len(communityCards) < 3 {
		return newError(ErrInvalidArgument, "there should be no less than 3 community cards")
	}
	if len(communityCards) > 5 {
		return newError(ErrInvalidArgument, "there should be no more than 5 community cards")
	}
	if numOfOpponents < 1 {
		return newError(ErrInvalidArgument, "there should be at least 1 opponent")
	}
	if numberOfSimulatedGames < 100 {
		return newError(ErrInvalidArgument, "number of simulated games is less than 100")
	}

	allCards := slices.Clone(communityCards)
	allCards = append(allCards, playerHoleCards.First, playerHoleCards.Second)
	return validateUniqueCards(allCards, "either playerHoleCards or communityCards arguments have duplicate cards")
}

func validatePreflopSimulation(playerHoleCards HoleCards, numOfOpponents int, numberOfSimulatedGames int) error {
	if numOfOpponents < 1 {
		return newError(ErrInvalidArgument, "there should be at least 1 opponent")
	}
	if numberOfSimulatedGames < 100 {
		return newError(ErrInvalidArgument, "number of simulated games is less than 100")
	}
	return validateHoleCards(playerHoleCards, "playerHoleCards")
}

func validatePreflopLookup(playerHoleCards HoleCards, numOfOpponents int) error {
	if numOfOpponents < 1 {
		return newError(ErrInvalidArgument, "there should be at least 1 opponent")
	}
	return validateHoleCards(playerHoleCards, "playerHoleCards")
}

func validateHoleCards(playerHoleCards HoleCards, paramName string) error {
	if playerHoleCards.First.Equal(playerHoleCards.Second) {
		return newError(ErrDuplicateCards, fmt.Sprintf("%s argument has duplicate cards", paramName))
	}
	return nil
}

func validateUniqueCardsAndNoLowAces(cards []Card, duplicateMessage string) error {
	if err := validateUniqueCards(cards, duplicateMessage); err != nil {
		return err
	}

	for _, card := range cards {
		if card.Rank == 1 {
			return newError(ErrLowAces, "when instantiating Ace cards use rank 14 not 1")
		}
	}

	return nil
}

func validateUniqueCards(cards []Card, duplicateMessage string) error {
	seen := make(map[[2]int]struct{}, len(cards))
	for _, card := range cards {
		key := [2]int{card.Rank, int(card.Suit)}
		if _, ok := seen[key]; ok {
			return newError(ErrDuplicateCards, duplicateMessage)
		}
		seen[key] = struct{}{}
	}
	return nil
}
